File: 2020/j5.cpp
#include <vector>
#include <queue>
#include <string>
#include <iostream>
#include <unordered_map>
using namespace std;

bool searched[1001][1001];
int main()
{
    cin.sync_with_stdio(0);
    cin.tie(0);

    unordered_map<int, vector<pair<int, int>>> vals;
    int m, n;
    cin >> m >> n;

    for (int row = 1; row <= m; ++row)
    {
        for (int col = 1; col <= n; ++col)
        {
            int val;
            cin >> val;
            searched[row][col] = false;
            vals[val].push_back({row, col});
        }
    }

    queue<pair<int, int>> q;
    q.push({m, n});
    while (!q.empty())
    {
        pair<int, int> cur = q.front();
        q.pop();
        int product = cur.first * cur.second;
        if (vals.find(product) != vals.end())
        {
            vector<pair<int, int>> &pairs = vals[product];
            for (pair<int, int> i : pairs)
            {
                if (i.first == 1 && i.second == 1)
                {
                    cout << "yes" << '\n';
                    return 0;
                }
                else if (!searched[i.first][i.second])
                {
                    searched[i.first][i.second] = true;
                    q.push(i);
                }
            }
        }
    }
    cout << "no" << '\n';
    return 0;
}

File: 2020/j5.py
"""
CCC '21 J5/S2 - Escape Room
Find this problem at:
https://dmoj.ca/problem/ccc20s2
"""

# !Note this solution TLE on the last test batch.
# The issue can likely be solved by using a large pre-filled
# visited array instead of the current visited set.
# Refer to j5.cpp for final solution.

import sys

room = []
products = {}
x, y = int(sys.stdin.readline()), int(sys.stdin.readline())

for row in range(1, x + 1):
    new_row = []
    col = 0
    for val in map(int, sys.stdin.readline().split()):
        col += 1
        new_row.append(val)
        product = row * col
        if product in products:
            products[product].append((row - 1, col - 1))
        else:
            products[product] = [(row - 1, col - 1)]
    room.append(new_row)

visited = set()
queue = [(0, 0)]

while queue:
    curr = queue.pop(0)
    product = room[curr[0]][curr[1]]
    for next in products.get(product, []):
        if next == (x - 1, y - 1):
            print("yes")
            break
        if next not in visited:
            visited.add(next)
            queue.append(next)
    else:
        continue
    break
else:
    print("no")
